### PR TITLE
PLANNER-2143 Validate config XML with a namespace

### PR DESCRIPTION
**Pull request by rsynek**
### JIRA
https://issues.redhat.com/browse/PLANNER-2143

### Referenced pull requests
https://github.com/kiegroup/optaplanner-website/pull/280

### Checklist
- [x] Documentation updated if applicable.
- [x] Upgrade recipe provided if applicable.

TODO: documentation, upgrade recipe and release notes to wrap up the entire XSD work.

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>full downstream build</b> please add comment: <b>Jenkins run fdb</b>
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/GenericJaxbIO.java [hunk 1]
```diff
@@ -129,119 +155,141 @@ public T readAndValidate(Reader reader, Schema schema) {
...
-
-        InputSource xmlInputSource = new InputSource(reader);
         try {
-            // Parse the XML to feed its content into the UnmarshallerHandler.
-            namespaceOverridingXmlFilter.parse(xmlInputSource);
-        } catch (IOException | SAXException e) {
-            final String errMessage = String.format(ERR_MSG_READ_OVERRIDE_NAMESPACE, rootClass.getName(),
-                    Arrays.toString(elementNamespaceOverrides));
-            throw new OptaPlannerXmlSerializationException(errMessage, e);
+            return builder.parse(new InputSource(nonNullReader));
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/GenericJaxbIO.java:233]: Shouldn't the `reader` be in `try (...)` ?

**Review by triceo — COMMENTED** [review]
So far so good.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/GenericJaxbIO.java (hunk 1 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/GenericJaxbIO.java:233]: Good question. The reader is created outside the method, usually from a stream (which is created inside a try-with-resources).

But it doesn't hurt to make sure it's closed even here.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/config/PlannerBenchmarkConfig.java [hunk 2]
```diff
@@ -158,8 +158,8 @@ public static PlannerBenchmarkConfig createFromXmlResource(String benchmarkConfi
             }
             return createFromXmlInputStream(in, classLoader);
         } catch (OptaPlannerXmlSerializationException e) {
-            throw new IllegalArgumentException("Unmarshalling of benchmarkConfigResource (" + benchmarkConfigResource
-                    + ") fails.", e);
+            throw new IllegalArgumentException(
```
[the remark this answers is not in the document]
Comment (rsynek) [reply optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/config/PlannerBenchmarkConfig.java:161]: Thanks, however, I am keeping the second formatting change as now it fits a single line.

**Review by ge0ffrey — APPROVED** [review]
This fixes a Windows issue I guess? :)

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/io/PlannerBenchmarkConfigIO.java [hunk 3]
```diff
@@ -24,16 +24,36 @@
...
+                    ElementNamespaceOverride.of(SolverConfig.XML_ELEMENT_NAME, SolverConfig.XML_NAMESPACE));
+        } else if (rootElementNamespace == null || rootElementNamespace.isEmpty()) {
+            // If not, add the missing namespace to maintain backward compatibility.
+            return genericJaxbIO.readOverridingNamespace(document,
+                    ElementNamespaceOverride.of(PlannerBenchmarkConfig.XML_ELEMENT_NAME, PlannerBenchmarkConfig.XML_NAMESPACE),
+                    ElementNamespaceOverride.of(SolverConfig.XML_ELEMENT_NAME, SolverConfig.XML_NAMESPACE));
+        } else { // If there is an unexpected namespace, fail fast.
+            String errMsg = String.format("The <%s/> element belongs to a different namespace (%s) than expected (%s).",
+                    PlannerBenchmarkConfig.XML_ELEMENT_NAME, rootElementNamespace, PlannerBenchmarkConfig.XML_NAMESPACE);
+            throw new IllegalArgumentException(errMsg);
```
Comment (ge0ffrey) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/io/PlannerBenchmarkConfigIO.java:55]: Nitpick: in other cases in this PR, that errMsg is called errorMessage.
Personally, I'd just inline that String every time into the call to the *Exception constructor.

**Comment by ge0ffrey** [comment]
+1 so far so good

**Comment by rsynek** [comment]
The coverage gap is all about fail fasts and exception translation. I added tests for the `SolverFactory` and the `PlannerBenchmarkFactory` that make sure a user will get both the root cause and the path to the resource or file that could not have been parsed.
What we are missing is tests for every possible exception that can ever raise from the XML handling code.

File: optaplanner-docs/src/main/asciidoc/BenchmarkingAndTweaking/BenchmarkingAndTweaking-chapter.adoc [hunk 4]
```diff
@@ -95,7 +95,8 @@ A benchmark configuration XML file looks like this:
 [source,xml,options="nowrap"]
 ----
 <?xml version="1.0" encoding="UTF-8"?>
-<plannerBenchmark>
+<plannerBenchmark xmlns="https://www.optaplanner.org/xsd/benchmark" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="https://www.optaplanner.org/xsd/benchmark https://www.optaplanner.org/xsd/benchmark/benchmark.xsd">
```
Comment (triceo) [inline optaplanner-docs/src/main/asciidoc/BenchmarkingAndTweaking/BenchmarkingAndTweaking-chapter.adoc:99]: This is kind of insane. I understand that we want to nudge people towards XSD, and that we want them to copy-paste this, but had I seen this for the first time, I'd run scared.

Solutions? I don't have any. :-(

Comment (triceo) [inline optaplanner-docs/src/main/asciidoc/BenchmarkingAndTweaking/BenchmarkingAndTweaking-chapter.adoc:99]: This is kind of insane. I understand that we want to nudge people towards XSD, and that we want them to copy-paste this, but had I seen this for the first time, I'd run scared.

Solutions? I don't have any. :-(

**Review by triceo — APPROVED** [review]
Just one concern, and no solution.

File: optaplanner-docs/src/main/asciidoc/BenchmarkingAndTweaking/BenchmarkingAndTweaking-chapter.adoc (hunk 4 above)
Comment (rsynek) [reply optaplanner-docs/src/main/asciidoc/BenchmarkingAndTweaking/BenchmarkingAndTweaking-chapter.adoc:99]: One option I see is taking a shortcut:
- we change any reference to <plannerBenchmark/> and <solver/> to `<plannerBenchmark xmlns="https://www.optaplanner.org/xsd/benchmark">` or `<solver xmlns="https://www.optaplanner.org/xsd/solver">` without the XSD location
- we show the full root element declaration in the Configuration chapter.

As a result, if someone copies such an incomplete declaration into a project, the IDE won't be able to download the XSD from the website. In such a situation IntelliJ IDEA looks for the XSD on the classpath.

Comment (triceo) [reply optaplanner-docs/src/main/asciidoc/BenchmarkingAndTweaking/BenchmarkingAndTweaking-chapter.adoc:99]: That could work.

Comment (rsynek) [reply optaplanner-docs/src/main/asciidoc/BenchmarkingAndTweaking/BenchmarkingAndTweaking-chapter.adoc:99]: @ge0ffrey what is your opinion on this? Which way would you prefer?

Comment (ge0ffrey) [reply optaplanner-docs/src/main/asciidoc/BenchmarkingAndTweaking/BenchmarkingAndTweaking-chapter.adoc:99]: We need the full declaration as in this PR, I believe. People will just copy paste it, they do it for pom.xml headings too, I am not to worried about it looking scary.

Why not just the "xmlns"? Then the IDE doesn't know where to download the xsd from.
Why not without any xmlns? Then they have no code completion nor xml validation while typing.

Comment (ge0ffrey) [reply optaplanner-docs/src/main/asciidoc/BenchmarkingAndTweaking/BenchmarkingAndTweaking-chapter.adoc:99]: One could argue that the XSD in xml spec has made that unnecessarily verbose - and I'd agree - but it is what it is.

Comment (rsynek) [reply optaplanner-docs/src/main/asciidoc/BenchmarkingAndTweaking/BenchmarkingAndTweaking-chapter.adoc:99]: A summary:
- other IDE than IDEA might not be able to handle the incomplete declaration gracefully
- this is how the declaration looks like "by the book"

These points outweigh the concern regarding the rather long and ugly declaration.

**Pull request merged**